python/cuml/dask/common/input_utils.py

```diff
@@ -0,0 +1,314 @@
+        to_map = dist_arr
+        parts = list(map(delayed, to_map))
+
+    elif isinstance(dask_obj, Iterable):
+        parts = [arr.to_delayed().ravel() for arr in dask_obj]
+        to_map = zip(*parts)
+        parts = list(map(delayed, to_map))
+
+    else:
+        raise TypeError("Unsupported dask_obj type: " + type(dask_obj))
+
+    parts = client.compute(parts)
+    yield wait(parts)
+
+    key_to_part_dict = dict([(str(part.key), part) for part in parts])
+    who_has = yield client.who_has(parts)
+
+    worker_map = {}  # Map from part -> worker
+    for key, workers in who_has.items():
+        worker = first(workers)
+        worker_map[key_to_part_dict[key]] = worker
+
+    worker_to_parts = []
+    for part in parts:
+        worker = worker_map[part]
+        worker_to_parts.append((worker, part))
+
+    yield wait(worker_to_parts)
+    raise gen.Return(worker_to_parts)
+
+
+@gen.coroutine
+def _extract_colocated_partitions(X_ddf, y_ddf, client=None):
+    """
+    Given Dask cuDF input X and y, return an OrderedDict mapping
+    'worker -> [list of futures] of X and y' with the enforced
+     co-locality for each partition in ddf.
+
+    :param X_ddf: Dask.dataframe
+    :param y_ddf: Dask.dataframe
+    :param client: dask.distributed.Client
+    """
+
+    client = default_client() if client is None else client
+
+    if isinstance(X_ddf, dcDataFrame):
+        data_parts = X_ddf.to_delayed()
+        label_parts = y_ddf.to_delayed()
+
+    else:
+        data_arr = X_ddf.to_delayed().ravel()
+        to_map = data_arr
+        data_parts = list(map(delayed, to_map))
+        label_arr = y_ddf.to_delayed().ravel()
+        to_map2 = label_arr
+        label_parts = list(map(delayed, to_map2))
+        # data_parts = X_ddf.to_delayed().ravel()
+        # label_parts = y_ddf.to_delayed().ravel()
+
+    parts = list(map(delayed, zip(data_parts, label_parts)))
+    parts = client.compute(parts)
+    yield wait(parts)
+
+    key_to_part_dict = dict([(part.key, part) for part in parts])
+    who_has = yield client.scheduler.who_has(
+        keys=[part.key for part in parts]
+    )
+
+    worker_map = defaultdict(list)
+    for key, workers in who_has.items():
+        worker_map[first(workers)].append(key_to_part_dict[key])
+
+    return worker_map
+
+
+def _workers_to_parts(futures):
+    """
+    Builds an ordered dict mapping each worker to their list
+    of parts
+    :param futures: list of (worker, part) tuples
+    :return:
+    """
+    w_to_p_map = OrderedDict()
+    for w, p in futures:
+        if w not in w_to_p_map:
+            w_to_p_map[w] = []
+        w_to_p_map[w].append(p)
+    return w_to_p_map
+
+
+def _get_ary_meta(ary):
+    return ary.shape, ary.dtype
+
+
+def _get_rows_from_colocated_tuple(obj):
+    return obj[0].shape[0]
+
+
+def _get_rows_from_single_obj(obj):
+    return obj.shape[0]
+
+
+def to_dask_cupy(futures, dtype=None, shapes=None, client=None, verbose=False):
+
+    wait(futures)
```

Comment:
Do we need to wait on this here? could we just let _get_ary_meta get started and wait for that? Maybe not possible... just trying to brainstorm some ways to make this more dask-graphy.